Improve `README.md`, add `CONTRIBUTING.md`, and more! (#152)
TheBevyFlock/bevy_cli: bevy_lint/README.md

<div class = "rustdoc-hidden">

# `bevy_lint`

`bevy_lint` is a custom linter for the [Bevy game engine](https://bevyengine.org), similar to [Clippy](https://doc.rust-lang.org/stable/clippy).

</div>

<div class="rustdoc-alert rustdoc-alert-warning">

> **Warning**
>
> This is an unofficial community project, hacked upon by the Bevy CLI working group until it is eventually upstreamed into the main [Bevy Engine organization]. Pardon our rough edges!

</div>

[Bevy Engine organization]: https://github.com/bevyengine

## Installation

> `bevy_lint` uses [`#![feature(rustc_private)]`](https://doc.rust-lang.org/nightly/unstable-book/language-features/rustc-private.html) to link to `rustc` crates. As such, it requires a specific nightly toolchain to be installed.

Reviewer: I think that explaining or using synonyms for both "link to" and "toolchain" would go a long way to making this less intimidating.

> `bevy_lint` uses [`#![feature(rustc_private)]`](https://doc.rust-lang.org/nightly/unstable-book/language-features/rustc-private.html) to link to `rustc` crates. As such, it requires a specific nightly toolchain to be installed.

Reviewer: This would really benefit from an explanation / justification of why this is necessary.

### Bleeding edge

You can install `bevy_lint` directly from the Git repository ([TheBevyFlock/bevy_cli](https://github.com/TheBevyFlock/bevy_cli)) to try out new and unstable features!

First, you must install the toolchain and components described by [`rust-toolchain.toml`](https://github.com/TheBevyFlock/bevy_cli/blob/main/rust-toolchain.toml) using [Rustup]. As of the time of writing (October 17th, 2024), the command may look like this:

```bash

> $ rustup toolchain install nightly-2024-10-03 \

Author: The version is specified in [`rust-toolchain.toml`](https://github.com/TheBevyFlock/bevy_cli/blob/main/rust-toolchain.toml). A user would replace `nightly-2024-10-03` with the value in the `channel = "..."` field.

This is a bit rough, I'll admit. How can I rephrase this to be more clear?

> $ rustup toolchain install nightly-2024-10-03 \

Reviewer: I think I would do:

First, you must install the exact toolchain expected by the version of the tool that you're using. You can determine the correct version by checking [`rust-toolchain.toml`]. For example, if you were using the commit `HASH`, you should look at...

      --component rustc-dev \
      --component llvm-tools-preview
```

Please be aware that you must keep this toolchain installed for `bevy_lint` to function[^0].

Next, install the actual linter from Git:

```bash
$ cargo +nightly-2024-10-03 install \
      --git https://github.com/TheBevyFlock/bevy_cli.git \
      --locked \
      bevy_lint
```

<div class="rustdoc-alert rustdoc-alert-important">

> **Important**
>
> Make sure to specify the correct nightly toolchain (such as `nightly-2024-10-03`) when running `cargo install`.

</div>

[Rustup]: https://rustup.rs

[^0]: `bevy_lint` interfaces with `rustc` to setup custom lints. It does not bundle all of `rustc` into the executable, though, and instead dynamically links to `librustc_driver.so` at runtime. `librustc_driver.so` is installed with the toolchain, so removing the toolchain will cause `bevy_lint` to fail to link.

## Usage

`bevy_lint` has the same API as `cargo check`:

```bash
$ bevy_lint --help
```

If you have the Bevy CLI installed, the linter as available as the `lint` subcommand:

```bash
$ bevy lint --help
```

> ### Configuring Lints

Reviewer: Can we configure these at the workspace level? Regardless of the answer, it should be listed here.

Author: It is not supported, added to the table!

> Toggling lints and lint groups requires the nightly `register_tool` feature. Even if your project uses stable Rust, you can still use this feature by detecting the `--cfg bevy_lint` flag:

Reviewer: This framing isn't clear to me. Start with a clear description of what the user is trying to do :)

```rust,ignore
// When `--cfg bevy_lint` is passed, enable the nightly `register_tool` feature and register
// `bevy`.
#![cfg_attr(bevy_lint, feature(register_tool), register_tool(bevy))]
```

`bevy_lint` checks your project with a nightly toolchain and automatically passes `--cfg bevy_lint`. Make sure to add it to the list of known `--cfg` flags in `Cargo.toml`:

```toml
[lints.rust]
unexpected_cfgs = { level = "warn", check-cfg = ["cfg(bevy_lint)"] }
```

> You can now toggle lints and lint groups throughout the crate, as long as they are also behind `#[cfg_attr(...)]`:

Reviewer: ```suggestion
You can now toggle lints and lint groups throughout your project, as long as they are also behind `#[cfg_attr(...)]`:
```

Author: I worry about saying this because you need to call `#![register_tool(bevy)]` for each crate root that you use the `bevy` tool namespace. This means all libraries, binaries, examples, and integration tests. Let me try thinking of a better way to communicate this...

```rust,ignore
#![cfg_attr(bevy_lint, warn(bevy::pedantic))]

#[cfg_attr(bevy_lint, deny(bevy::panicking_world_methods))]
fn my_critical_system(world: &mut World) {
    // ...
}
```

> If you do not register `bevy` as a tool, `#[allow(bevy::lint_name)]` and related attributes will fail to compile.

Reviewer: This needs to be moved up: motivation first.

|-|-|-|
|`#[allow(...)]` and related|✅|Must be behind `#[cfg_attr(bevy_lint, ...)]` on stable Rust|
|`[lints.bevy]` in `Cargo.toml`|⚠️|(Nightly only because `#[register_tool(bevy)]` must not be behind `#[cfg_attr(bevy_lint, ...)]`)|
|`RUSTFLAGS="-A bevy::lint"`|❌|`RUSTFLAGS` applies to dependencies, but they do not `#[register_tool(bevy)]`|

<div class="rustdoc-alert rustdoc-alert-tip">

> **Tip**
>
> If your project uses nightly Rust by default, you can forego the `#![cfg_attr(...)]` and write `#![feature(register_tool)]` and `#![register_tool(bevy)]` directly. This will let you configure lints using the `[lints.bevy]` table in `Cargo.toml`.

</div>

## Compatibility

|`bevy_lint` Version|Rust Version|Bevy Version|
|-|-|-|
|0.1.0-dev|rustc 1.83.0-nightly (18b1161ec 2024-10-02)|0.14|

## License

The Bevy Linter is licensed under either of

- Apache License, Version 2.0 ([`LICENSE-APACHE`](https://github.com/TheBevyFlock/bevy_cli/blob/main/LICENSE-APACHE) or <http://www.apache.org/licenses/LICENSE-2.0>)
- MIT license ([`LICENSE-MIT`](https://github.com/TheBevyFlock/bevy_cli/blob/main/LICENSE-MIT) or <http://opensource.org/licenses/MIT>)

at your option.

> ## Contributing

Reviewer 2: Can we include some more help for contributors here? At the very least, I think it would be helpful to explain how to run the UI tests.
